### ISSUE TITLE
Long Output is chunked when the process is busy

### ISSUE THREAD

**Issue by orgads** [git-for-windows/git#3706]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

64-bit. I build it myself, but it reproduces also with official releases. See [here](https://sourceware.org/pipermail/cygwin/2022-February/250916.html).

```
$ git --version --build-options

git version 2.35.1.windows.2.5.gae9e853979
cpu: x86_64
built from commit: ae9e853979a9c939d18a75ddb5daae183c622023
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19043.1466]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: WinSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Rebase
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Enabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash, but it reproduces also on cmd.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
mkdir -p chunked
cd chunked
git init
mkdir -p abc/def/ghi/jkl def
cd abc/def/ghi/jkl
touch foo
git add foo
seq 1 1000 | xargs touch
cd ../../../../def
seq 1 30000 | xargs touch
cd ..
time git clean -dfx
```
 - What did you expect to occur after running these commands?

The output should be synchronized. While deleting the long `def` directory, all the previous files should be written.

 - What actually happened instead?

The output hangs in the middle of a line, and continues from the same place after several seconds.

Initially reported [here](https://github.com/git-for-windows/git/issues/3657#issuecomment-1041189952).

**Comment by dscho** [comment]
Does this also happen in a recent Cygwin snapshot (or for that matter, in MSYS2)?

**Comment by orgads** [comment]
I use the snapshot that you provided in #3657. Is there a more recent snapshot?

**Comment by dscho** [comment]
No, but I'm trying to help route this to the correct project. If the same issues occur in Cygwin, that would give me a bit breathing room in Git for Windows.

**Comment by rimrul** [comment]
> Bash, but it reproduces also on cmd.
>
> time git clean -dfx

Does this reproduce in `cmd` without `time`?

**Comment by orgads** [comment]
> If the same issues occur in Cygwin, that would give me a bit breathing room in Git for Windows.

It reproduces on Cygwin only with Git for Windows. With Cygwin Git it doesn't reproduce.

**Comment by dscho** [comment]
> Edit: Yes, it does. I don't think there's any Msys2 runtime involved in that case.

Hmm. I don't recall any recent changes in Git for Windows. Does `git clean` write to `stderr`? If so, I would expect the pipe and the thread in `compat/winansi.c` to be the culprit...

**Comment by rimrul** [comment]
`git clean` seeems to just use `printf()` for that message, so `stdout`.

I think we're experiencing the situation described in https://github.com/adda-team/adda/issues/277.

`mingw-w64` seems to default to full buffering of `stdout`, and we'd need to either use `setvbuf(stdout,NULL,_IONBF,0)` or `fflush(stdout)` to get around this.

**Comment by orgads** [comment]
Pushed https://github.com/git/git/pull/1447
